quda_interface.c

```diff
@@ -1114,34 +1181,112 @@ void _setMultigridParam(QudaMultigridParam* mg_param) {
-      // cases divisible by 3, otherwise we use 2
-      // if blocking is not possible, we use block size 1
-      if( extent == 1 ) mg_param->geo_block_size[i][j] = 1;
-      else mg_param->geo_block_size[i][j] = (i == 0) ? 4 : ( (extent % 3 == 0) ? 3 : even_block_size );
+  if (Quda_multigrid_input.nlevel > QUDA_MAX_MG_LEVEL ){
+    if (g_cart_id == 0) {printf("Error in number of levels in QUDA MG %d\n", Quda_multigrid_input.nlevel );
+                         printf("It exceeds the maximal number of levels\n");
+                         fflush(stdout);
+                         exit(1);
+                        }
+  }
+  mg_param->n_level = Quda_multigrid_input.nlevel;
+  if (Quda_multigrid_input.blocksize[0][0] != 0){
+      unsigned int extent=1;
+      for (int i=0; i<mg_param->n_level; i++){
+        extent*=Quda_multigrid_input.blocksize[0][i];
+      }
+      if (extent != LX){
+        if (g_cart_id == 0) { printf("Error in input for block sizes in direction x\n");
+                              fflush(stdout);
+                              exit(1);
+                            }
+      }
+      for (int i=0; i<mg_param->n_level; i++){
+        mg_param->geo_block_size[i][0]=Quda_multigrid_input.blocksize[0][i];
+      }
+      extent=1;
+      for (int i=0; i<mg_param->n_level; i++){
+        extent*=Quda_multigrid_input.blocksize[1][i];
+      }
+      if (extent != LY){
+        if (g_cart_id == 0) { printf("Error in input for block sizes in direction y\n");
+                              fflush(stdout);
+                              exit(1);
+                            }
+      }
+      for (int i=0; i<mg_param->n_level; i++){
+        mg_param->geo_block_size[i][1]=Quda_multigrid_input.blocksize[1][i];
+      }
+      extent=1;
+      for (int i=0; i<mg_param->n_level; i++){
+        extent*=Quda_multigrid_input.blocksize[2][i];
+      }
+      if (extent != LZ){
+        if (g_cart_id == 0) { printf("Error in input for block sizes in direction z\n");
+                              fflush(stdout);
+                              exit(1);
+                            }
+      }
+      for (int i=0; i<mg_param->n_level; i++){
+        mg_param->geo_block_size[i][2]=Quda_multigrid_input.blocksize[2][i];
+      }
+      extent=1;
+      for (int i=0; i<mg_param->n_level; i++){
+        extent*=Quda_multigrid_input.blocksize[3][i];
+      }
+      if (extent != T){
+        if (g_cart_id == 0) { printf("Error in input for block sizes in direction T\n");
+                              fflush(stdout);
+                              exit(1);
+                            }
+      }
+      for (int i=0; i<mg_param->n_level; i++){
+        mg_param->geo_block_size[i][3]=Quda_multigrid_input.blocksize[3][i];
+      }
+  }
+  else{
+    for (int i=0; i<mg_param->n_level; i++) {
+      for (int j=0; j<QUDA_MAX_DIM; j++) {
+
+        unsigned int extent = (j == 4) ? T : LX;
```

Comment:
This was a hack at the time, in principle, this should choose between LX, LY, LZ and T and required. There's an array for that somewhere in geometry